packages/core/lib/Client.ts

```diff
@@ -0,0 +1,136 @@
+import * as Adapter from './Adapter';
+import { Event, IBreadcrumb, IUser, LogLevel } from './Interfaces';
+import * as Context from './Interfaces/Context';
+import { DSN } from './Interfaces/DSN';
+import { IOptions } from './Options';
+import { SentryError } from './Sentry';
+
+// TODO: Add breadcrumbs
+// TODO: Add context handling tags, extra, user
+export class Client {
+  public readonly dsn: DSN;
+  private _adapter: Adapter.IAdapter;
+  private _context: Context.IContext;
+
+  constructor(
+    dsn: string,
+    public options: IOptions = {
+      logLevel: LogLevel.Error,
+      maxBreadcrumbs: 100,
+    }
+  ) {
+    this.dsn = new DSN(dsn);
+    this._context = Context.getDefaultContext();
+    return this;
+  }
+
+  public getContext() {
+    // TODO: check for cyclic objects
+    return JSON.parse(JSON.stringify(this._context));
+  }
+
+  private get adapter() {
+    if (!this._adapter) {
+      throw new SentryError('No adapter in use, please call .use(<Adapter>)');
+    }
+    return this._adapter;
+  }
+
+  public getAdapter<A extends Adapter.IAdapter>(): A {
+    return this.adapter as A;
+  }
+
+  /**
+   * Register a Adapter on the Core
+   * @param adapter
+   * @param options
+   */
+  public use<A extends Adapter.IAdapter, O extends {}>(
+    adapter: { new (client: Client, options?: O): A },
+    options?: O
+  ): Client {
+    if (this._adapter) {
+      // TODO: implement unregister
+      throw new RangeError(
+        'There is already a Adapter registered, call unregister() to remove current adapter'
+      );
+    }
+    this._adapter = new adapter(this, options);
+    return this;
+  }
+
+  public async captureException(exception: Error) {
+    return this.send(await this.adapter.captureException(exception));
+  }
+
+  public async captureMessage(message: string) {
+    return this.send(await this.adapter.captureMessage(message));
+  }
+
+  public captureBreadcrumb(crumb: IBreadcrumb) {
+    return this.adapter.captureBreadcrumb(crumb);
+  }
+
+  public install() {
+    return this.adapter.install();
```

Comment:
**Disclaimer:** I assume that `install()` performs some actions to initialize the adapter and make it ready for event submission etc...

I'm a bit worried about this function. The return value is not specified explicitly here, but apparently it is `Promise<boolean>`. This indicates two things:
 * The installation might take a while in which the client is not necessarily ready
 * It can fail in two ways: By resolving `false` or by rejecting some `Error`. In both cases, the use should not be allowed to use the client anymore

In most of your examples, you however do stuff like this:

```ts
Sentry.create('https://53039209a22b4ec1bcc296a3c9fdecd6@sentry.io/4291')
  .use(Browser)
  .install();
Sentry.getSharedClient().setTagsContext({ cordova: true });
Sentry.getSharedClient().captureBreadcrumb({ message: 'HOHOHOHO' });
Sentry.getSharedClient().captureMessage('PICKLE RIIIICK!');
Sentry.getSharedClient().captureException(new Error('YOYOYOY'));
```

So this means that by the time you start configuring the shared client, it's not necessarily installed and you might have an unhandled promise rejection. Instead, I would propose to resolve to the client instance here and maybe even remove the `getSharedClient()` method:

```ts
public install() {
  if (await this.adapter.install()) {
    return this;
  }

  throw new Error('Client installation failed');
}
```

That way, the user is more encouraged to wait for the result:

```ts
const client = await Sentry.create('https://53039209a22b4ec1bcc296a3c9fdecd6@sentry.io/4291')
  .use(Browser)
  .install();
client.setTagsContext({ cordova: true });
client.captureBreadcrumb({ message: 'HOHOHOHO' });
client.captureMessage('PICKLE RIIIICK!');
client.captureException(new Error('YOYOYOY'));
```

One more thing, if you really want to be safe, introduce an internal flag in the client whether installation has succeeded, before allowing methods like `captureException`.